browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage))
+    return;
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  std::string domain = URLToStorageDomain(navigation_handle->GetURL());
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (domain == previous_domain)
+    return;
+
+  ClearEphemeralStorageIfNecessary();
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance = content::SiteInstance::CreateForURL(
+      browser_context, navigation_handle->GetURL());
+  auto* partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance.get());
+
+  std::string local_partition_id = domain + "/ephemeral-local-storage";
+  content::SessionStorageNamespaceMap::const_iterator it =
+      local_storage_namespace_map().find(local_partition_id);
+  if (it == local_storage_namespace_map().end()) {
+    auto local_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, local_partition_id);
+    local_storage_namespace_map()[local_partition_id] =
+        std::move(local_storage_namespace);
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
+        std::move(session_storage_namespace);
+  }
+}
+
+void EphemeralStorageTabHelper::WebContentsDestroyed() {
+  ClearEphemeralStorageIfNecessary();
+}
+
+bool EphemeralStorageTabHelper::IsDifferentWebContentsWithMatchingStorageDomain(
+    WebContents* other_web_contents,
+    const std::string& storage_domain) {
+  DCHECK(other_web_contents);
+  if (web_contents() == other_web_contents)
+    return false;
+
+  std::string other_storage_domain =
+      URLToStorageDomain(other_web_contents->GetLastCommittedURL());
+  return storage_domain == other_storage_domain;
+}
+
+bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
+    const std::string& storage_domain) {
+#if defined(OS_ANDROID)
+  for (auto it = TabModelList::begin(); it != TabModelList::end(); ++it) {
+    TabModel* tab_model = *it;
+    if (tab_model->GetProfile() == profile) {
+      for (int i = 0; i < tab_model->GetTabCount(); i++) {
+        TabAndroid* tab = tab_model->GetTabAt(i);
+        WebContents* web_contents = tab->web_contents();
+        if (IsDifferentWebContentsWithMatchingStorageDomain(web_contents,
+                                                            storage_domain))
+          return true;
+      }
+    }
+  }
+#else
+  for (Browser* browser : *BrowserList::GetInstance()) {
```

Comment:
on shutdown we will do this for each destroyed web_contents, thus it looks like O(N^2) ops on UI thread. I'd suggest to reconsider having a data structure for this

Reply:
Hrm. That's an interesting point and I think it would also be true even if we used another data structure here. It seems a solution could be to special-case shutting down.

Reply:
I'd say special-case window closing

Reply:
I have spent a bit of time thinking about this and I think that your initial idea of creating a new data structure makes a lot of sense. I have gone ahead and done that. The new design, which I will upload soon does this:
  * Adds a new reference-counted data structure `PerTLDEphemeralStorage` which manages all per-TLD ephemeral storage (currently) only localStorage.
  * Each `EphemeralStorageTabHelper` keeps a reference to this data structure which is fetched from a map of `WeakPtr<PerTLDEphemeralStorage>` when the domain changes.
  * When the last tab helper using this per-TLD storage is freed or switches domains, reference counting ensures that the storage is cleaned up.

I think this new design has a bunch of nice benefits:
  * It's less code.
  * It removes a dependency on the UI bits of the browser that were causing a dependency cycle. We can turn on `check_includes` again without any issues.
  * It's more efficient, because we don't have to iterate through tabs or look into a map. The reference counting code is highly-optimized.
   * We only keep one map instead of two. Session storage is now fully managed by the tab helper.